### PR TITLE
Add exception type and update xml docs. 

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
- Adds new `InvalidImageContentException` type.
- Documents exceptions and adds guards on all loading mechanisms.
- Documents exceptions and adds guards on all saving mechanisms.
- Minor cleanup and performance tweaks.

Fixes. #1110

### CODE REVIEW

File: src/ImageSharp/Image.FromBytes.cs [hunk 1]
```diff
@@ -65,13 +69,16 @@ public static IImageFormat DetectFormat(Configuration configuration, byte[] data
...
         /// <param name="data">The byte array containing encoded image data to read the header from.</param>
         /// <param name="format">The format type of the decoded image.</param>
-        /// <exception cref="NotSupportedException">Thrown if the stream is not readable.</exception>
+        /// <exception cref="ArgumentNullException">The configuration is null.</exception>
+        /// <exception cref="ArgumentNullException">The data is null.</exception>
+        /// <exception cref="NotSupportedException">The data is not readable.</exception>
         /// <returns>
         /// The <see cref="IImageInfo"/> or null if suitable info detector is not found.
         /// </returns>
         public static IImageInfo Identify(Configuration configuration, byte[] data, out IImageFormat format)
```
Comment (antonfirsov) [inline src/ImageSharp/Image.FromBytes.cs:78]: Can decoders throw `InvalidImageContentException` in `Identify`?

Comment (JimBobSquarePants) [reply src/ImageSharp/Image.FromBytes.cs:78]: Yeah, they can hit a bad value in the image header.

**Comment by JimBobSquarePants** [comment]
> Great job! Will have another one in a few days.

Quick as possible if you can. I'm hoping to ship in less than a week.

**Comment by brianpopow** [comment]
@JimBobSquarePants there are still some places left with a `TODO: use InvalidImageContentException here, if we decide to define it`

For example https://github.com/SixLabors/ImageSharp/blob/5639babe23a438958f5f59db86c332728f15707f/src/ImageSharp/Formats/Gif/GifDecoder.cs#L43.

Was this intentional or should we change those, too?

**Comment by JimBobSquarePants** [comment]
@brianpopow I missed that. Yeah we should update.

**Comment by brianpopow** [comment]
> @brianpopow I missed that. Yeah we should update.

@JimBobSquarePants ok i will do that

**Pull request merged**